### PR TITLE
Adding formatted output for trivy and dockle

### PR DESCRIPTION
**Pull request by shigupt202**
Printing formatted output for trivy and dockle. The output looks like this now:
image

### CODE REVIEW

**Comment by thesattiraju** [comment]
Have you tried this? https://www.npmjs.com/package/table

**Comment by thesattiraju** [comment]
Update the image in description

File: src/main.ts [hunk 1]
```diff
@@ -109,9 +112,15 @@ async function runDockle(): Promise<number> {
 async function run(): Promise<void> {
     whitelistHandler.init();
     const trivyStatus = await runTrivy();
+    if (trivyStatus === trivyHelper.TRIVY_EXIT_CODE) {
+        trivyHelper.printFormattedOutput();
+    }
```
Comment (thesattiraju) [inline src/main.ts:121]: You should add some info in `else` too.

File: src/main.ts [hunk 2]
```diff
@@ -109,9 +112,15 @@ async function runDockle(): Promise<number> {
...
     let dockleStatus: number;
     if (inputHelper.isCisChecksEnabled()) {
         dockleStatus = await runDockle();
+        if (dockleStatus === dockleHelper.DOCKLE_EXIT_CODE) {
+            dockleHelper.printFormattedOutput();
+        }
```
Comment (thesattiraju) [inline src/main.ts:132]: Same here

File: src/trivyHelper.ts [hunk 3]
```diff
@@ -128,3 +129,44 @@ function getTrivyDownloadUrl(trivyVersion: string): string {
...
+    let titles = ["VULNERABILITY ID", "PACKAGE NAME", "SEVERITY", "DESCRIPTION"];
+    rows.push(titles);
+    trivyOutputJson.forEach(ele => {
+        if (ele && ele["Vulnerabilities"]) {
+            ele["Vulnerabilities"].forEach((cve: any) => {
+                let row = [];
+                row.push(cve["VulnerabilityID"]);
+                row.push(cve["PkgName"]);
+                row.push(cve["Severity"]);
+                row.push(cve["Description"]);
```
Comment (thesattiraju) [inline src/trivyHelper.ts]: Get. the keys from some constant

File: src/trivyHelper.ts [hunk 4]
```diff
@@ -128,3 +129,44 @@ function getTrivyDownloadUrl(trivyVersion: string): string {
             throw new Error(util.format("Container scanning is not supported on %s currently", curOS));
     }
 }
+
+export function printFormattedOutput() {
+    const trivyOutputJson = getTrivyOutput();
+    let rows = [];
+    let titles = ["VULNERABILITY ID", "PACKAGE NAME", "SEVERITY", "DESCRIPTION"];
```
Comment (thesattiraju) [inline src/trivyHelper.ts]: Get titles from constants

File: src/trivyHelper.ts [hunk 5]
```diff
@@ -128,3 +129,44 @@ function getTrivyDownloadUrl(trivyVersion: string): string {
...
+            width: 25,
+            wrapWord: true
+          },
+          3: {
+            width: 65,
+            wrapWord: true
+          }
+        }
+      };
+    console.log(table.table(rows, config));
```
Comment (thesattiraju) [inline src/trivyHelper.ts]: This code is duplicate. Move it to utils

File: src/main.ts [hunk 6]
```diff
@@ -76,30 +76,33 @@ async function getDockleEnvVariables(): Promise<{ [key: string]: string }> {
...
     const imageName = inputHelper.imageName;
 
     const dockleOptions: ExecOptions = {
         env: dockleEnv,
-        ignoreReturnCode: true
+        ignoreReturnCode: true,
+        silent: true
     };
-
+    console.log("Scanning for CIS and best practice violations...");
```
Comment (thesattiraju) [inline src/main.ts:105]: Add TRIVY logs to debug as well.

File: src/main.ts [hunk 7]
```diff
@@ -76,30 +76,33 @@ async function getDockleEnvVariables(): Promise<{ [key: string]: string }> {
...
     const docklePath = await dockleHelper.getDockle();
+    core.debug(util.format("Dockle executable found at path ", docklePath));
     const dockleEnv = await getDockleEnvVariables();
     const imageName = inputHelper.imageName;
 
     const dockleOptions: ExecOptions = {
         env: dockleEnv,
-        ignoreReturnCode: true
+        ignoreReturnCode: true,
+        silent: true
```
Comment (thesattiraju) [inline src/main.ts:103]: You'll have to add log parsing logic to find failures of scan too.

**Review by thesattiraju — CHANGES_REQUESTED** [review]

**Comment by thesattiraju** [comment]
I understand it's extra work. Could you move the node_modules to a different PR? 1300+ files are just too many to scroll. Let that go in right away. This shouldn't cause issues for your testing purposes

**Comment by ds-ms** [comment]
I understand it's extra work. Could you move the node_modules to a different PR? 1300+ files are just too many to scroll. Let that go in right away. This shouldn't cause issues for your testing purposes

File: src/trivyHelper.ts (hunk 3 above)
Comment (shigupt202) [reply src/trivyHelper.ts]: Done

File: src/trivyHelper.ts (hunk 5 above)
Comment (shigupt202) [reply src/trivyHelper.ts]: Moved

File: src/main.ts (hunk 6 above)
Comment (shigupt202) [reply src/main.ts:105]: It's already there in runTrivy function.

File: src/utils.ts [hunk 8]
```diff
@@ -53,4 +53,29 @@ function getCheckText(trivyStatus: number, dockleStatus: number): string {
     }
 
     return text;
+}
+
+export function getConfigForTable(col1_width: number, col2_width: number, col3_width: number, col4_width: number): any {
```
Comment (ajinkya599) [inline src/utils.ts]: low priority: Should we stick to camelCase for function parameters?

File: src/utils.ts [hunk 9]
```diff
@@ -53,4 +53,29 @@ function getCheckText(trivyStatus: number, dockleStatus: number): string {
...
+    let config = {
```
Comment (ajinkya599) [inline src/utils.ts:59]: Do you want to keep this dynamic? (Pass an array columnWidths and return the config accordingly. That way addition/deletion of columns would be simple in future.

File: src/trivyHelper.ts [hunk 10]
```diff
@@ -4,12 +4,23 @@ import * as fs from 'fs';
...
-const semver = require('semver');
+import * as table from 'table';
+import * as semver from 'semver';
+import * as utils from './utils';
 
 export const TRIVY_EXIT_CODE = 5;
 const stableTrivyVersion = "0.5.2";
 const trivyLatestReleaseUrl = "https://api.github.com/repos/aquasecurity/trivy/releases/latest";
 const trivyToolName = "trivy";
+const VULNERABILITIES = "Vulnerabilities";
```
Comment (ajinkya599) [inline src/trivyHelper.ts]: Prefix these constants with KEY, like `KEY_VULNERABILITIES`, `KEY_VULNERABILITY_ID` and so on?

File: src/dockleHelper.ts [hunk 11]
```diff
@@ -4,13 +4,22 @@ import * as fs from 'fs';
...
-import * as inputHelper from './inputHelper';
-const semver = require('semver');
+import * as table from 'table';
+import * as semver from 'semver';
 
 export const DOCKLE_EXIT_CODE = 5;
 const stableDockleVersion = "0.2.4";
 const dockleLatestReleaseUrl = "https://api.github.com/repos/goodwithtech/dockle/releases/latest";
 const dockleToolName = "dockle";
+const DETAILS = "details";
```
Comment (ajinkya599) [inline src/dockleHelper.ts]: Prefix these constants with `KEY_`?

**Comment by ajinkya599** [comment]
LGTM. Left a few comments.
@shigupt202 - Can you please link the issues that will be addressed with this PR?

File: src/dockleHelper.ts (hunk 11 above)
Comment (shigupt202) [reply src/dockleHelper.ts]: Done

File: src/trivyHelper.ts (hunk 10 above)
Comment (shigupt202) [reply src/trivyHelper.ts]: Done

File: src/utils.ts (hunk 9 above)
Comment (shigupt202) [reply src/utils.ts:59]: Replaced with an array

File: src/utils.ts (hunk 8 above)
Comment (shigupt202) [reply src/utils.ts]: Not using them anymore

File: src/main.ts (hunk 7 above)
Comment (shigupt202) [reply src/main.ts:103]: I think we can add those changes in a separate PR as it needs some more investigation. Will be tracking the issue [here](https://github.com/shigupt202/local-container-scan/issues/11).

File: src/trivyHelper.ts (hunk 4 above)
Comment (shigupt202) [reply src/trivyHelper.ts]: Done

File: src/main.ts (hunk 1 above)
Comment (shigupt202) [reply src/main.ts:121]: Added

File: src/main.ts (hunk 2 above)
Comment (shigupt202) [reply src/main.ts:132]: Added

**Review by thesattiraju — APPROVED** [review]

**Pull request merged**